Feat/add ai semantic caching plugin (#7713)
Kong/docs.konghq.com: app/_hub/kong-inc/ai-semantic-cache/overview/_index.md

Together, these caching methods enhance the efficiency and relevance of AI responses, making interactions faster and more contextually accurate.

### Configuration parameters

You can configure the following required parameters for the AI Semantic Cache plugin:

* `vectordb`: Defines the vector database strategy and settings.
* `embeddings`: Defines the embeddings provider, model, and configuration
* `cache_control`: Enables or disables the `cache-control` header handling
* `storage_ttl`: Sets the `time-to-live` for cached responses
* `exact_caching`: Enables or disables exact caching
* `message_countback`: Specifies how many of the user's recent messages will be vectorized for querying the vector database. The entire response is still cached, this setting only determines how many past messages are turned into embeddings.
* `ignore_system_prompts`: Ignores system messages (`role = system`) in the chat history when sending the embeddings (to generate vectors) if set to `true`. 
* `ignore_assistant_prompts`: Ignores assistant messages (`role = assistant`) in the chat history when sending the embeddings (to generate vectors) if set to `true`. 

> * `stop_on_failure`: Stops processing if an error occurs, otherwise continues

Reviewer: You don't need this here at all. This is what's defined in the configuration reference page (/hub/kong-inc/ai-semantic-cache/configuration/), with a lot more detail + nested parameters there.

---
nav_title: Overview
---

## What is semantic caching?

> Semantic caching enhances data retrieval efficiency by focusing on the meaning or context of queries rather than just exact matches. It stores responses based on the underlying intent and semantic similarities between different queries and can then retrieve those cached queries when a similar request is made. For example, if a user asks, "how to integrate our API with a mobile app" and later asks "what are the steps for connecting our API to a smartphone application?", the LLM recognizes that both queries are seeking similar information about the steps to integrate an API with an app. It can then retrieve and reuse previously cached responses if they are contextually relevant, even if they are phrased differently. By leveraging semantic caching, the LLM can reduce redundant processing and speeds up the response times.

Reviewer 2: ```suggestion
Semantic caching enhances data retrieval efficiency by focusing on the meaning or context of queries rather than just exact matches. It stores responses based on the underlying intent and semantic similarities between different queries and can then retrieve those cached queries when a similar request is made.
```

Reviewer 2: I split it up so it doesn't look like  a wall of text

> When a new request is made, the system can retrieve and reuse previously cached responses if they are contextually relevant, even if the phrasing differs. This method reduces redundant processing, speeds up response times, and ensures that answers are more relevant to the user’s intent, ultimately improving overall system performance and user experience. 

Reviewer 2: ```suggestion
When a new request is made, the system can retrieve and reuse previously cached responses if they are contextually relevant, even if the phrasing is different. This method reduces redundant processing, speeds up response times, and ensures that answers are more relevant to the user’s intent, ultimately improving overall system performance and user experience. 

For example, if a user asks, "how to integrate our API with a mobile app" and later asks, "what are the steps for connecting our API to a smartphone application?", the system understands that both questions are asking for the same information. It can then retrieve and reuse previously cached responses, even if the wording is different. This approach reduces processing time and speeds up responses.
```
I stuck the example here so it fit with the first paragraph. I simplified the way it was written a bit too.

> By using the AI Semantic Cache plugin, the plugin can handle the caching for you. You can also configure the vector database, caching mechanism, as well as embeddings.

Reviewer 2: ```suggestion

```

> The AI Semantic Cache plugin might not be the best fit for you if following apply to you:
> * Storing semantic vectors or embeddings and performing similarity searches can require significant storage and computational resources. If you have limited hardware or budget constraints, this might be a concern.
> * If your application data isn't inherently semantic or exact matches are sufficient, the benefits of semantic caching might be minimal. Simple keyword-based or traditional caching might be more appropriate and efficient.

Reviewer 2: ```suggestion
The AI Semantic Cache plugin may not be ideal for you if:

* If you have limited hardware or budget, storing semantic vectors and running similarity searches need a lot of storage and computing power, which could be an issue.
* If your data doesn’t rely on semantics, or exact matches work fine, semantic caching may offer little benefit, and traditional or keyword-based caching might be more efficient.
```

## How it works

> {% mermaid %}

Reviewer 2: This diagram is not very easy to read. Is there a better way to display this, and if not do we need the diagram?

Reviewer 2: Maybe it's two diagrams?

        User[User]
    end

    subgraph APIGateway["API Gateway"]
        Kong[{{site.base_gateway}}]
    end

    subgraph AISemanticCache["AI Semantic Cache"]
        Plugin[AI Semantic Cache Plugin]
    end

    subgraph DataStorageRedis["Data Storage (Redis)"]
        VectorDB[Vector Database Role]
        Cache[Cache Role]
    end

    User -->|Request| Kong
    Kong -->|Forward Request| Plugin
    Plugin -->|Check Cache| Cache
    Cache -->|Cache Hit| Plugin
    Cache -->|Cache Miss| VectorDB
    VectorDB -->|Retrieve Embedding| Plugin
    Plugin -->|Generate Response| Kong
    Kong -->|Send Response| User
{% endmermaid %}
<!--vale on-->

> Figure 2: This diagram shows how the AI Semantic Cache plugin works with a vector database, cache, and {{site.base_gateway}}.

### Vector databases

A vector database can be used to store vector embeddings, or numerical representations, of data items. For example, a response would be converted to a numerical representation and stored in the vector database so that it can compare new requests against the stored vectors to find relevant cached items.

Currently, the AI Semantic Cache plugin supports Redis as vector databases.

> For more information, see [Deploy and manage a vector database](/hub/kong-inc/ai-semantic-cache/vector-database/).

Reviewer 2: Where is this doc?

### Cache management

With the AI Semantic Cache plugin, you can configure a cache of your choice to store the responses from the LLM.

> For more information, see [Cache management](/hub/kong-inc/ai-semantic-cache/cache-management/).

Reviewer 2: Where is this doc?

* **Semantic Caching:** This approach is more flexible and involves storing responses based on the meaning or intent behind the queries. Instead of relying on exact matches, the system can understand and reuse information that is conceptually similar. For instance, if a user asks about "Italian restaurants in New York City" and later about "New York City Italian cuisine," semantic caching can help provide relevant information based on their related meanings.

Together, these caching methods enhance the efficiency and relevance of AI responses, making interactions faster and more contextually accurate.

### Configuration parameters

You can configure the following required parameters for the AI Semantic Cache plugin:

* `vectordb`: Specifies the vector database strategy and settings
* `embeddings`: Defines the embeddings provider, model, and configuration
* `cache_control`: Enables or disables Cache-Control header handling
* `storage_ttl`: Sets the time-to-live for cached responses
* `exact_caching`: Enables or disables exact caching

> * `message_countback`: Specifies how many messages to consider for context

Reviewer 3: @AntoineJac Is this how many messages the plugin will go through to find a semantic match before determining that there is no match?

* `ignore_system_prompts`: Ignores system messages if set to `true`

> * `ignore_assistant_prompts`: Ignores assistant messages if set to `true`

Reviewer 3: @AntoineJac Could you explain what these do? I'm not sure what a system or assistant message is in this context

> * `vectordb`: Specifies the vector database strategy and settings

Reviewer 2: ```suggestion
* `vectordb`: Defines the vector database strategy and settings.
```

> * `cache_control`: Enables or disables Cache-Control header handling

Reviewer 2: ```suggestion
* `cache_control`: Enables or disables the`cache-control` header handling
```

> * `storage_ttl`: Sets the time-to-live for cached responses

Reviewer 2: ```suggestion
* `storage_ttl`: Sets the `time-to-live` for cached responses
```

> * `message_countback`: Specifies how many messages in the incoming array from the user will be vectorized. The whole response will still be cached. This only counts how many messages back in the chat history to turn into embeddings and query to the vector database.

Reviewer 2: ```suggestion
* `message_countback`: Specifies how many of the user's recent messages will be vectorized for querying the vector database. The entire response is still cached, this setting only determines how many past messages are turned into embeddings.
```
